hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapFetchEntriesMessageTask.java

```diff
@@ -53,7 +54,7 @@ protected ClientMessage encodeResponse(Object response) {
         }
         MapEntriesWithCursor mapEntriesWithCursor = (MapEntriesWithCursor) response;
         return MapFetchEntriesCodec.encodeResponse(mapEntriesWithCursor.getNextTableIndexToReadFrom(),
-                mapEntriesWithCursor.getBatch());
+                                                   (List<Map.Entry<Data, Data>>)(Object)mapEntriesWithCursor.getBatch());
```

Comment:
I believe this will create a class cast exception when being used from the client. We're going to have to release this changed - https://github.com/hazelcast/hazelcast-client-protocol/blob/1.8.z/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/template/ResponseTemplate.java#L219-L225.
cc @asimarslan @sancar

Reply:
The casting here looks confusing but it will not cause any cast exception. So no need for any protocol change.
`getBatch`  methods always returns `List<Map.Entry<Data, Data>>` , it is just not clear from the method signature. 

I think the problem arises from usage of the generics on AbstractCursor. We are using the generic for both the stored object on construction, and also for the object that we read. Their types are actually different. I also don't see a strong reason for AbstractConstructor to know the type. Child classes can keep the reference themselves and override read,write methods so that AbstractConstructor does not know the type.

Reply:
I've modified the `AbstractMapPartitionIteratorTest#test_HasNext_Returns_True_On_NonEmptyPartition` test to use `OBJECT` in-memory format and to add `Immutable` objects as values to the map. This forces the issue and I've been able to get the exception:
```
java.lang.ClassCastException: com.hazelcast.client.map.AbstractMapPartitionIteratorTest$Lala cannot be cast to com.hazelcast.nio.serialization.Data

	at com.hazelcast.client.impl.protocol.codec.MapFetchEntriesCodec$ResponseParameters.calculateDataSize(MapFetchEntriesCodec.java:107)
	at com.hazelcast.client.impl.protocol.codec.MapFetchEntriesCodec.encodeResponse(MapFetchEntriesCodec.java:119)
	at com.hazelcast.client.impl.protocol.task.map.MapFetchEntriesMessageTask.encodeResponse(MapFetchEntriesMessageTask.java:56)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.sendResponse(AbstractMessageTask.java:195)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.onResponse(AbstractPartitionMessageTask.java:89)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:256)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.execute(AbstractPartitionMessageTask.java:78)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.unblock(AbstractInvocationFuture.java:247)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.andThen(AbstractInvocationFuture.java:223)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processMessage(AbstractPartitionMessageTask.java:69)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:129)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:109)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:163)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:159)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:127)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:110)
```

This is because previously, `MapEntriesWithCursor` could only contain a list of serialized keys and values. The `MapFetchEntriesCodec#encodeResponse` method expects these to be `Data` to be able to calculate the data size and prepare the response. But now, due to the fact that we want to avoid having extra copies on reads, the `MapEntriesWithCursor` can also contain references to actual user objects in some cases, which is why it contains a list of `Map.Entry<Data, Object>`. In some cases, the value will be `Data` and in other cases it will be the user object. But now `MapFetchEntriesCodec#encodeResponse` fails to prepare the response in cases where the value is not serialised. The same issue is in the EE repo.

Reply:
I don't  know why but I was thinking that AbstractCursor will always come from remote. 
In that case, we need to convert to proper type before passing to encodeResponse. 
Probably in `MapFetchEntriesMessageTask.encodeResponse`  code.

Reply:
Hi mmedanjak,
This is Rajeev from Adobe. I was the original author of these changes. A little bit of context around this patch:

- This patch is intended and tailor made for our use case(s) and in no way meant to be a general purpose, comprehensive all-encompassing feature covering all use cases and apis. I did try to address this holistically but quickly realized that it would be a much larger effort and outside of our concern. 

- As described in the design doc, we are addressing a very narrow subset of our use cases. The whole intent was to minimize redundant copy on read and thus minimizing memory usage.

- You are right about the classcastexception that would happen when the iterator api is accessed from the clientproxy. I had realized this side effect of the patch. That would mean not being able to use the ClientMapProxy::iterator api (with object format on) which is anyways not exposed on the IMap interface (and we dont use direct Apis on the proxy). 

- So overall, yes the patch breaks an api (with object format) which we dont use. (even if we were to use that api, in that topology we would use the binary format) as a trade off against optimization of memory usage. We would like to accept and work within these caveats and proceed with the patch after addressing the other feedback.

- Re: benchmark
We havent done any benchmark. We did add test to confirm the patch functionality works and COR is disabled for local access of immutable types. Intuitively & naively, the performance should improve for the targeted use case and remain the same otherwise.

Reply:
The fix is quite simple actually here as @sancar pointed out. The collection will need to be serialised either way as we're transmitting it from the member back to the client. So, you can simply fix this by going over the collection and serialising the values.

Reply:
Thanks @mmedenjak @sancar Yes we will fix this issue.

Reply:
Looks good, thanks!